### ISSUE TITLE
-Wexit-time-destructors diagnoses usage in discarded statements

### ISSUE THREAD

**Issue by pkasting** [llvm/llvm-project#68651]
The following code generates a warning when compiled with `-Wexit-time-destructors`, even though no exit-time destructor is generated:
```
struct T { ~T(); };
struct S {
  T t;
};

void foo() {
  if constexpr(false) {
    static S s;
  }
}
```

Online repro: https://godbolt.org/z/n66e93f45

It seems like this warning should only be evaluated after discarded statements from a `constexpr if` are removed, or else needs to take the discard state into account?

**Comment by shafik** [comment]
So the diagnostic is `warn_exit_time_destructor` and we emit it here:

https://github.com/llvm/llvm-project/blob/feea5db01360b477b8cf2df03abfa9fc986633d5/clang/lib/Sema/SemaDeclCXX.cpp#L16119-L16121

and we are emitting the diagnostic from `Sema::EmitCurrentDiagnostic(...)`:

https://github.com/llvm/llvm-project/blob/feea5db01360b477b8cf2df03abfa9fc986633d5/clang/lib/Sema/Sema.cpp#L1611-L1615

I wonder if we need a check for `ExprEvalContexts.back().isDiscardedStatementContext()` someplace here.

CC @AaronBallman @zygoloid @cor3ntin

**Comment by cor3ntin** [comment]
@shafik I think we should use `DiagRuntimeBehavior` for `warn_exit_time_destructor`, this seem to be the cleanest way to fix that

**Comment by AaronBallman** [comment]
Normally we would use `DiagRuntimeBehavior()` but that relies on us building a CFG to determine if the code is reachable. The code which emits this is looking for uses that include dstructors of globals for which we aren't currently able to build a CFG around. So this is the correct suggestion, but it's going to be more involved to implement than it sounds like.

One idea I've had on how to implement this is whether we should have the CFG model flow control for globals as-if they were in a function context. e.g., construct a CFG that notionally looks like:
```
void __start(int argc, char *argv[], char *environ[]) {
  global_init_1();
  global_init_2();
  main(argc, argv, environ);
  at_exit_handlers();  
}
```
but I've not had the chance to play with this idea too much.

**Comment by shafik** [comment]
Perhaps relevant, [CWG2893](https://www.open-std.org/jtc1/sc22/wg21/docs/cwg_closed.html#2893)
